docs/source/data/fetching-data.mdx

```diff
@@ -1,6 +1,224 @@
+
+<MultiCodeBlock>
+
+```ts title="index.ts"
+//highlight-start
+interface ContextValue {
+  dataSources: {
+    dogsDB: DogsDataSource;
+    catsApi: CatsAPI;
+  };
+  token: string
+}
+//highlight-end
+
+const server = new ApolloServer<ContextValue>({
+  typeDefs,
+  resolvers,
+});
+
+const { url } = await startStandaloneServer(server, {
+  context: async () => {
+    const { cache } = server; // highlight-line
+    const token = req.headers.token;
+    return {
+      // We create new instances of our data sources with each request.
+      // We can pass in our server's cache, contextValue, or any other
+      // info our data sources require. 
+      // highlight-start
+      dataSources: {
+        dogsDB: new DogsDataSource({ cache, contextValue: token }),
+        catsApi: new CatsAPI({ cache }),
+      },
+      //highlight-end
+      token
+    };
+  },
+});
+
+console.log(`🚀  Server ready at ${url}`);
+```
+
+</MultiCodeBlock>
+
+<!-- TODO(AS4) add link to context article once exists  -->
+Apollo Server calls [the `context` initialization](./resolvers/#the-context-argument) function for _every incoming operation_. This means:
+- For every operation, `context` returns an _object_ containing new instances of your data source classes (in this case, `DogsDataSource` and `CatsAPI`).
+- The **`context` function should create a new instance of each data source class for each operation.** You might accidentally combine results from multiple operations if multiple operations share a single data source instance.
+
+Your resolvers can then access your data sources from the shared `contextValue` object and use them to fetch data:
+
+```ts title="resolvers.ts"
+const resolvers = {
+  Query: {
+    dog: async (_, { id }, { dataSources }) => {
+      return dataSources.dogsDB.getDog(id);
+    },
+    popularDogs: async (_, __, { dataSources }) => {
+      return dataSources.dogsDB.getMostLikedDogs();
+    },
+    bigCats: async (_, __, { dataSources }) => {
+      return dataSources.catsApi.getCats({ size: 10 });
+    },
+  },
+};
+```
+
+## Open-source implementations
+
+[In Apollo Server 3](/apollo-server/v3/data/data-sources), you'd subclass the `DataSource` abstract class for each of your data sources. In Apollo Server 4, we recommend [making a custom class for each data source](#creating-data-source-classes), with each class best suited for that particular source of data.
+
+> ⚠️ **Note**: The community built each data source package below for use with Apollo Server 3. [As shown below](#using-datasource-subclasses), you can still use these packages in Apollo Server 4 with a bit of extra setup.
+
+The below data source implementations extend the generic [`DataSource` abstract class](https://github.com/apollographql/apollo-server/blob/main/packages/apollo-datasource/src/index.ts), from the deprecated `apollo-datasource` package. Subclasses of `DataSource` define the logic required to communicate with a particular store or API.
```

Comment:
```suggestion
The below data source implementations extend the legacy [`DataSource` abstract class](https://github.com/apollographql/apollo-server/blob/main/packages/apollo-datasource/src/index.ts), from the deprecated `apollo-datasource` package. Subclasses of `DataSource` define the logic required to communicate with a particular store or API. They may still be useful in Apollo Server 4, though you will have to manually call the `initialize` method on them after creating the object.
```

alternatively we could have a table of "useful data source classes for Apollo Server 4" and another table of "Apollo Server 3 data source classes that we hope will be updated for AS4 soon".

Or another organization we could do is:

- have a section that lists (modern) data source classes.  link to, say, RESTDataSource for now

- have a *separate* section talking about legacy Apollo Server 3 data sources; this one can have the text about how AS3 used to have this extra thing, and you can probably still use AS3 data sources with AS4 if you call `.initialize` yourself (but ideally the projects will update to the simpler AS4 API), and list them

... oh ok that is sort of what you're doing already, but I'd just separate "here are new things" from "here's a list of old things and the extra thing you have to do in order to use them even though their docs probably don't tell you to"

Reply:
So I do have a Note _right_ about this paragraph saying that they'll need to do some more work to get old  these subclasses working:

> > ⚠️ **Note**: The community built each data source package below for use with Apollo Server 3.  [As shown below](#using-datasource-subclasses), you can still use these packages in Apollo Server 4 with a bit of extra setup.

So I don't feel like we need the extra sentence you suggested if that sounds good to you.

For organization: we won't have more data source classes in the future, right?  We expect folks to make their data source classes—so having a separate section about "modern" classes feels a bit weird if it'll only be `RESTDataSource`?  I do agree that I could do a better job of separating out these older implementations, though. I could change the header of this section to "Apollo Server 3 Open-source implementations" if that'd feel like a good way to mark it as "old stuff"?

Reply:
I mean this comes back to the general issue that this is a page documenting a concept that doesn't really exist any more... Like if we're going to have a table of AS3-style classes then maybe we would want to continue having tables in the future.  But also basically anything on npm could be a data source... The only thing that really makes it AS-specific is taking our cache class as an option...

Reply:
Resolving this with the Modern/Legacy sections